test/e2e/etcd_failure.go

```diff
@@ -0,0 +1,171 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/client"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/fields"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/labels"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/util/wait"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+const etcdFailureDuration = 20 * time.Second
+
+type etcdFailureCase struct {
+	c  *client.Client
+	ns string
+}
+
+var _ = Describe("Etcd failure", func() {
+	defer GinkgoRecover()
+
+	var c *client.Client
+	var ns string
+
+	BeforeEach(func() {
+		By("preparing namespace")
+
+		ns = "e2e-test-" + string(util.NewUUID())
+		var err error
+
+		c, err = loadClient()
+		Expect(err).NotTo(HaveOccurred())
+
+		Logf("creating new namespace")
+		_, err = createNamespaceIfDoesNotExist(c, ns)
+		Expect(err).NotTo(HaveOccurred())
+
+		Logf("creating replication controller before failure")
+		Expect(RunRC(RCConfig{
+			Client:    c,
+			Name:      "baz",
+			Namespace: ns,
+			Image:     "nginx",
+			Replicas:  1,
+		})).NotTo(HaveOccurred())
+	})
+
+	It("should recover from network partition with master", func() {
+		var dropEtcdPackets sync.Once
+		testEtcdFailure(etcdFailureCase{c, ns}, func() {
+			dropEtcdPackets.Do(func() {
+				masterExec("sudo iptables -A INPUT -p tcp --destination-port 4001 -j DROP")
+			})
+		}, func() {
+			masterExec("sudo iptables -D INPUT -p tcp --destination-port 4001 -j DROP")
+		})
+	})
+
+	It("should recover from SIGKILL", func() {
+		testEtcdFailure(etcdFailureCase{c, ns}, func() {
+			masterExec("pgrep etcd | xargs -I {} sudo kill -9 {}")
+		}, func() { /* nothing */ })
+	})
+})
+
+func testEtcdFailure(efc etcdFailureCase, cause func(), fix func()) {
+	// This test requires SSH, so the provider check should be identical to
+	// there (the limiting factor is the implementation of util.go's
+	// getSigner(...)).
+	if !providerIs("gce", "gke") {
+		By(fmt.Sprintf("Skipping reboot test, which is not implemented for %s", testContext.Provider))
+		return
+	}
+
+	By("failing etcd")
+
+	stop := make(chan struct{}, 1)
+	go util.Until(func() {
+		defer GinkgoRecover()
+		cause()
+	}, 1*time.Second, stop)
+	time.Sleep(etcdFailureDuration)
+	stop <- struct{}{}
+	fix()
+
+	assertEtcdRecovers(efc)
+}
+
+func masterExec(cmd string) {
+	Logf("sshing into master %v to execute '%v'", getMasterHost(), cmd)
```

Comment:
How much logging do we like to have in our e2e tests? Do we use any leveled logging in e2e?